### ISSUE TITLE
Trigger Enhancement

### ISSUE THREAD

**Issue by Thiana** [legendblade/CraftingHarmonics#63]
Since you're adding the potion effects feature..

Could you also add the ability to trigger a minecraft command when the potion effect is removed? For example it could be used to say something to the player, or even apply/remove another set.

Possibly three triggers in the case of the potion effect, OnApply, OnRemove, OnExpire? 

It might be useful for other types of sets as well to create mutually exclusive sets. ie, auto-remove set B when set A or set C are applied. Or even have things like 'fatigue' sets auto apply after 'bonus' sets. ie, do something that triggers extreme mob shedding which when it wears off would then stop the mob shedding at all for a certain period.

Though this is probably out of scope for your vision of CH if you were aiming for a MineTweaker type mod?

**Comment by legendblade** [comment]
I think we're well beyond this being just a MineTweaker replacement by this point, and branching out like this is good (to an extent), because it helps the mod stand out.  :)

Let me sleep on this, and I'll see if I can come up with a decent solution.

**Comment by legendblade** [comment]
So, some thoughts on this:

- Add an optional 'duration' to sets, giving them the ability to automatically turn themselves off after a certain amount of time.
- Add a 'runCommand' operation, which allows you to specify a command to run when the set:
  - Is applied
  - Is removed
  - Expires
- Add a 'runPlayerCommand' operation, which allows you to specify a command to run on a specific player (or, y'know, globally, if you don't use '@p' in the command), with support for the matcher system, in case simple potion effects aren't good enough. :)
- Add a way to chain some of the drop/shed/potion effect code so that you can apply/remove potion effects or run a command when a certain drop / potion effect is triggered. (ie: mob sheds something and gets regen 1 for a few seconds; player harvests a block, gets poison 2 for a minute, etc).
  - Regular drops would have:
    - onDropped
  - Potion effects would have:
    - onApplied
    - onRemoved
    - onExpired
  - When implemented, mob spawns would probably have an 'onSpawned' callback as well
- Allow you to run a JavaScript function in place of simply calling a command operation for all of those.

Now, some of that won't be in the first iteration (I'm thinking this will be 1.4.2 through... 1.4-point-whatever), but, I think that should expand things nicely, and open up some more advanced functionality that we've been missing?

**Comment by legendblade** [comment]
@Thiana also... I think I somehow deleted your comment when I went to add mine... *facepalm*

**Comment by Thiana** [comment]
That looks good though in the case of runPlayerCommand I'd suggest additional events of 'onFailure' and 'onSuccess' since the set is an instant action it can either succeed or fail at the start. So it would have onApply, onFailure, onSuccess.

To clarify, onSuccess/onApply are not the same event. onApply would run regardless of the outcome of the operation matchers.

**Comment by Thiana** [comment]
Actually, for runPlayerCommand just have onFailure/onSuccess. If the modpack creator wants the onApply they can just run the same action for both onFailure/onSuccess.

**Comment by Thiana** [comment]
Another thing to consider. A better way might be to allow something like:`
onApply: { execute: [commands], apply-set: [sets], remove-set: [sets] } ` with the execute/apply-set/remove-set being optional operations. That removes the need to pass control to MC's command interpreter constantly for (potentially) long action chains.

**Comment by legendblade** [comment]
So, regarding onApply vs onSuccess/onFailure: if you're familiar with the jQuery JS framework, there are many times where it's useful to have shared code that always runs after the success/failure - so, maybe something like:

```
onSuccess: {
    execute: [
       {
          type: "runCommand",
          ...
       }
    ]
},
onFailure: function(error) {
   console.log("Something went wrong: " + error);
},
always: {
   removeSet; ["custom"]
}
```

Just woke up, so, kinda still working it all out in my brain.  :)

**Comment by Thiana** [comment]
I've used jQuery for some minor things, but not enough that I would say I'm familiar with it. 

Are you talking about having a generic `on<Event>` defined that is then applied to *all* sets?  If it's a global event action then you'd probably have to pass some data along to it for it to be useful. At least the set name that invoked the event? 

If you just want a generic action shared with many sets, then that could be simulated easily by having a `onApply: { applySet: ["generic-onapply"] }`? Where generic-onapply could be  a runPlayerCommand type set.

In terms of event naming, perhaps stick to an on<Event> convention? That way it's clear it's a reaction to the set action rather than a matcher to see if the set will be applied. ie, always -> onApply/onUse/etc.  To me `always` seems more like an always true matcher for testing ;)

**Comment by legendblade** [comment]
So, right now, with spending some time working on it today, I've got the following for potions:

```
{
    // Each effect has many of the same matchers as mob sheds
    what: "minecraft:luck",
    amplifier: 0, // The 'level' to add of the potion.  Defaults to 0. (in this case, 0 is Luck 1, and 1 is Luck 2)
    showParticles: false, // If effect particles should be shown while this is active.  Defaults to false; I hate potion particles.
    duration: 200,  // The duration in ticks of the effect (200 = 10 seconds).  If not specified, this defaults to the number of ticks between checks +1. 
                    // If you make this longer than the number of seconds, you can end up stacking effects
    cures: [], // A list of consumables that can cure this effect.  By default this will be nothing.
    cooldown: 200, // Make sure it's only applied every 200 ticks as well.
    // We also have event callbacks when these occur:
    onNew: [], // Called when a new potion is applied
    onExtended: [], // Called when the duration of an existing potion is extended
    onApplied: [ // Called every time the effect happens (after onNew/onExtended)
        { 
            // You can supply either an object...
            type: "runPlayerCommand",
            command: "/say @p You've got some extra luck.",
            name: "The Server" // For commands that show a name, you can set the name here.
        },
        function(entity, world) {
            // Or a function, which gets the entity, and the world.
            console.log(entity.getName() + " was blessed with some luck.");
        }
    ],
    onExpired: [], // Called when the duration expires without being cured
},
```

Right now, only the "runPlayerCommand" object is really all that useful, because there's not a lot being passed to the entity/world objects for the JS callback.  That said, you could still use the "runPlayerCommand" part to apply/remove sets, and do most anything else that a command can do.

Now, to the bad news... Forge doesn't provide (as far as I can tell) a way to check when the potion is just removed.  I can get around this (in a few ways), but that will require more work to implement.

So, that's what I've got so far, and will probably be... most of the extent of what will get released with 1.4.2.  For 1.4.3, I'll go ahead and add in the "runCommand" / "runPlayerCommand" operations (along with the matcher system for runPlayerCommand) as things you can do in a set instead of just as a callback, and add in support for matchers with the runPlayerCommand callback.  Plus probably make the actual arguments passed to the JavaScript functions a bit more useful.

Hopefully that's not too big of a disappointment? :-/

**Comment by Thiana** [comment]
Just to clarify onApplied... The event would be called each time the potion effect occurs ie, a heal in the case of Regen? If so, please please rename it to something else (onEffect maybe?). If I were reading the config file without knowing the events beforehand and didn't see onNew I'd assume onApplied would be the event triggered when the potion effect is first applied to the target.

Also, unless you've decided not to include applySet/removeSet triggered events then it would be better to keep onApplied/onRemoved for them. 

How do runPlayerCommand / runCommand differ?

Not a disappointment at all. Also, there is no rush for any of this :) Remember, I'm still waiting on a several core mods so I'm just working out progression/lore/worldgen and seeing what I can use to get the effects I want. When I decide on a progression point I then try and execute it with CH/BQ. If I can't, I come bother you :) If I can, I note the solution and move on until I encounter a new point that I dont have a solution for and repeat. You adding new features usually simplifies quite a few of my solutions, so I go back and update those :) 

To give you an idea, mods I must have are Intangible [you have to summon the gods before you can kill them] and EnderIO (or maybe NeoTech/BC) [for the mid-tier tech]. I'm also hoping for Mekanism [high tech], maybe Thaumcraft [mid magic] and Draconic Evolution (supreme tech/magic -- that will be the end game mod -- the player has mastered magic and tech and become as powerful as the gods they initially depended on), but those can wait. I wouldnt mind if I could find an 'overpowered' magic mod to match with Mekanism but I suspect those will all be post-Thaumcraft release.

Basically I want a pack on the scale of Regrowth but without the hand holding of HQM and with a large dash of fear for your life.

**Comment by legendblade** [comment]
Alright, so... my brain is a little dead, because I've been having to coremod so I could create the 'onRemoved' callback (which isn't the same as 'onCured', which... I'll have to further research) because there's no Forge event or way to hook into that... so, my responses may be a bit jumbled. :)

onApplied: My initial description was misleading and your instincts are correct.  It gets called anytime the potion is either applied (after onNew) or extended (after onExtended).  This is much the same as onRemoved (which may need a better name itself...) which is called after onExpired/onCured whenever the potion effect ends (onEnded? onPotionStoppedBeingAThing?)

I was intending to, initially, not include applySet/removeSet triggers, just leaving those for commands (/ch applySet -s ...) at the moment, if that's alright?

runCommand is intended for initial setup of things (ie: scoreboard add / etc) that go along with a set.  It will always be applied when a set is applied, and will have both a "apply" and "remove" command (so that you can run a command when the set is applied, and run a command when the set is removed).  

On the other hand, runPlayerCommand is intended to be more of an event in reaction to something (ie: hasScoreboard), which can target a specific player.  Thus why runPlayerCommand will support matchers, and runCommand will not.

That all said... your pack sounds pretty cool - I'll definitely have to give it a go when you get it all done... apparently right after I have to add Crafting Harmonics support for EnderIO, NeoTech, BC (Charset?), Mekanism, Thaumcraft, DE, etc... :laughing:

**Comment by legendblade** [comment]
The first bit of this is done as of commit b452dbf.  Will expand in future releases; moving this off to release 1.4.3 for future features.

**Comment by Thiana** [comment]
Have you considered doing a PR what you would need for onRemoved instead of core modding? The Forge team seems more open to accepting PRs lately.  

If by triggers you mean actions to replace `/ch applySet` and `/ch removeSet`. That is fine. I only brought up that idea as a potential optimization option. The part I wanted was the events but it looks like you're implementing that as onSetApplied=runCommand.apply and onSetRemoved=runCommand.remove.  

Actually, my current method for 'unlocking' higher mods is to gate the crafting recipes with materials that the gated mod makes itself. For example, an EIO Alloy Furnace would have a recipe that requires EIO alloys so to make the initial Alloy Furnace you would need a drop from a deity. So while I'd probably use that mod support if you added it, I don't actually need it with my current plans :) The ongoing difficulty comes not from gating every machine in the mods themselves, but the materials used to make them. They won't get access to real mob farming until high tech/magic (RFTools spawning/Soul Shards), and new dimensions (rftools dimensions) where they can quarry are even higher tech. Obviously it becomes easier as they get higher machines because progression gives better resource multiplication and possibly cheaper recipes for lower tiers.  

I'm also playing with the idea of allowing the player to trigger a 'retrogen' type effect for a small chunk of the world at a time so they have some alternatives to constantly running after mobs for resources. A sort of precursor to dimension creation in a way.

**Comment by legendblade** [comment]
So, have I considered it?  Yes.  However, for someone like me, it's a bit of a scary prospect.  I mean, it's Forge, and my code is (IMO) not "good enough" right now.  I do want to spend some time looking at it, and seeing if it's something that can (and should) be done, and, being that it's adding a new event... it'd probably help other people out.  And avoiding a coremod is always a good thing.  But it's... y'know... *contributing to Forge*, and that's still a scary thing. :)

Regarding the applySet/removeSet stuff... I think I've managed to get us both confused (mostly myself), so... I'll just say, right now, if you want to apply or remove a set based on an effect, then use the 'runPlayerCommand' callback, and we'll swing back around to this discussion when we start tackling actually improving sets (for instance: timed sets, exclusive sets, callbacks for onApplied/onRemoved, etc).

As far as adding support for the mods - the way I see it, I have to do it either way to stay... competitive with the other options available... unless I were to literally just give you the ability to run MineTweaker commands in a set (not sure if I even could, and... that seems... dangerous, anyway).

**Comment by legendblade** [comment]
Sorry, I suppose I should expand on a side thought as well: I'm trying to consider SMP as well as SSP when implementing these features.  From an SMP standpoint, it (generally) makes little sense to unlock a set in response to a player event, because that suddenly opens up that set for everyone, and, in most cases, the things that a set does are impossible to restrict only to a single player (crafting recipes, for instance, 99% of the time, can't be locked to only the player doing the crafting*).  That's definitely not to say that doing it in a SSP or cooperative SMP environment wouldn't be fine (packs like Crash Landing or Regrowth, for instance), but, for other packs (for instance, Infinity Evolved Skyblock in SMP) it wouldn't be.

> Technically, there are a few ways to deal with per-player crafting - some more invasive than others - as such, it's more accurate to say it's a complexity I don't want to get into just yet.

So, I guess the thing that I'm trying to get at is that... while I'm trying to make it possible to do these things - because I think it's a largely unexplored way of doing a progression style pack - I'm trying to be careful about how I do it, to try and make sure that it's viable for both SSP/co-op SMP as well as it is for a server.

Now, in no way am I saying that adding in the functionality you're asking for is going against that, I'm just trying to be cautious about implementing things that make it seem like you can apply a set directly to a player right now.  Most of the event callbacks that I have planned - see #72 for a better list of them - are generally centered around doing something to the specific entity that triggered the event.

... and I probably don't need to have said any of this.  :laughing:

**Comment by Thiana** [comment]
Well it was interesting :) But you're correct that my modpack is definitely SSP only. Having the world change dynamically would just become horribly complicated when more reacting to more than one person. Even a team wouldn't solve that problem since you could easily end up with contradictory sets being applied.

That said, the ability to limit effect to an area around a selector mentioned in #72 *might* allow some team based play but you could still easily break the pack. ie, if the members of the team end up on different tiers. Offhand I can't think of any way to avoid that and still keep the progression I want.

**Comment by legendblade** [comment]
Well, at least in regards to the contradictory sets... at the very least, I think that adding:
- Durations and cooldowns for sets
- A dependsOn: "otherSetName", prohibitedWith: "someOtherSet", etc
- Maybe some other things (including, at least, a removesSets: ["setThisConsumes"])

... might help?

**Comment by Thiana** [comment]
Actually, with one other change I think it could allow for team play.

Currently the min/maxScore have two params... name/value. Could you add an optional 3rd? The player to check against.

ie, allow isTier1=`maxScore: { name: "WorldTier", player: "TierTracker", value: 1 }`. Then I could use something like `dependsOn: "isTier1"` Add another scoreboard which tracks the number of players within a tier, then only advance to Tier2 when Tier1Players==0

**Comment by Thiana** [comment]
For the sake of simplification, please add an `isScore: { name: "foo", player: "bar", value: n }`. That way we dont need dependsOn: "set". More efficient to just check the scoreboard directly and isScore bypasses the need for both minScore/maxScore in the same set.

**Comment by legendblade** [comment]
See #78 - I think that should about cover that?
